File: ios/MyNoteCore/Sources/MyNoteCore/LocalStore.swift
import Foundation

/// What the sync engine needs from local persistence.
///
/// Async because the real implementation is a SwiftData `ModelActor` — a
/// `ModelContext` is not thread-safe, so it has to own its own isolation. Tests
/// use `InMemoryStore`, which satisfies the same contract.
public protocol LocalStore: Sendable {
    /// Highest `serverSeq` this device has durably applied.
    func cursor() async throws -> Int
    func setCursor(_ value: Int) async throws

    /// Local edits not yet acknowledged by the server, oldest first.
    func outbox(limit: Int) async throws -> [Change]
    func enqueue(_ change: Change) async throws

    /// Drop outbox entries the server has resolved — accepted or permanently
    /// rejected — identified by the exact version that was sent.
    ///
    /// Removal is by `(key, hlc)`, never by key alone. Both stores collapse a
    /// burst of typing onto one outbox row, so a keystroke landing *while a push
    /// is in flight* overwrites that row with a newer clock. Deleting by key
    /// would drop the newer edit without ever having sent it: silent, permanent
    /// note loss.
    func removeFromOutbox(_ sent: [Change]) async throws

    /// Newest clock this device has seen, across records and the outbox.
    ///
    /// Used to seed the clock at launch so it cannot restart behind its own
    /// previous edits after the system clock moves backwards.
    func newestHlc() async throws -> String?

    /// Wipe every local record, the outbox and the cursor.
    ///
    /// Called when the signed-in account changes: local rows carry no `uid`, so
    /// without this the previous account's queued edits would be pushed into the
    /// new one.
    func clearAll() async throws

    /// The clock on our copy of a record, or nil if we have never seen it.
    func currentHlc(_ entity: Entity, _ id: String) async throws -> String?

    /// Overwrite the local copy with the server's version.
    func applyRemote(_ change: Change) async throws
}

public struct ChangeKey: Hashable, Sendable {
    public let entity: Entity
    public let id: String
    public init(entity: Entity, id: String) {
        self.entity = entity
        self.id = id
    }
}

public extension Change {
    var key: ChangeKey { ChangeKey(entity: entity, id: id) }
}

/// Reference implementation used by tests and SwiftUI previews.
public actor InMemoryStore: LocalStore {
    private var records: [ChangeKey: Change] = [:]
    private var pending: [ChangeKey: Change] = [:]
    private var pendingOrder: [ChangeKey] = []
    private var cursorValue = 0

    public init() {}

    public func cursor() throws -> Int { cursorValue }

    public func setCursor(_ value: Int) throws { cursorValue = value }

    public func outbox(limit: Int) throws -> [Change] {
        pendingOrder.prefix(limit).compactMap { pending[$0] }
    }

    public func enqueue(_ change: Change) throws {
        records[change.key] = change
        if pending[change.key] == nil { pendingOrder.append(change.key) }
        pending[change.key] = change
    }

    public func removeFromOutbox(_ sent: [Change]) throws {
        for change in sent {
            // Only if this is still the version we pushed.
            guard pending[change.key]?.hlc == change.hlc else { continue }
            pending.removeValue(forKey: change.key)
            pendingOrder.removeAll { $0 == change.key }
        }
    }

    public func newestHlc() throws -> String? {
        (records.values.map(\.hlc) + pending.values.map(\.hlc)).max()
    }

    public func clearAll() throws {
        records.removeAll()
        pending.removeAll()
        pendingOrder.removeAll()
        cursorValue = 0
    }

    public func currentHlc(_ entity: Entity, _ id: String) throws -> String? {
        records[ChangeKey(entity: entity, id: id)]?.hlc
    }

    public func applyRemote(_ change: Change) throws {
        records[change.key] = change
    }

    // MARK: Test helpers

    public func record(_ entity: Entity, _ id: String) -> Change? {
        records[ChangeKey(entity: entity, id: id)]
    }

    public func allRecords() -> [Change] { Array(records.values) }
}

File: android/core/src/main/kotlin/io/mynote/core/LocalStore.kt
package io.mynote.core

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * What the sync engine needs from local persistence. The app backs this with
 * Room; tests use [InMemoryStore].
 */
interface LocalStore {
    /** Highest `serverSeq` this device has durably applied. */
    suspend fun cursor(): Int
    suspend fun setCursor(value: Int)

    /** Local edits not yet acknowledged by the server, oldest first. */
    suspend fun outbox(limit: Int): List<Change>
    suspend fun enqueue(change: Change)

    /** Drop outbox entries the server accepted (or permanently rejected). */
    suspend fun removeFromOutbox(keys: List<ChangeKey>)

    /** The clock on our copy of a record, or null if we have never seen it. */
    suspend fun currentHlc(entity: Entity, id: String): String?

    /** Overwrite the local copy with the server's version. */
    suspend fun applyRemote(change: Change)
}

/** Reference implementation used by tests. */
class InMemoryStore : LocalStore {
    private val mutex = Mutex()
    private val records = mutableMapOf<ChangeKey, Change>()
    private val pending = LinkedHashMap<ChangeKey, Change>()
    private var cursorValue = 0

    override suspend fun cursor(): Int = mutex.withLock { cursorValue }

    override suspend fun setCursor(value: Int) = mutex.withLock { cursorValue = value }

    override suspend fun outbox(limit: Int): List<Change> =
        mutex.withLock { pending.values.take(limit) }

    override suspend fun enqueue(change: Change) = mutex.withLock {
        records[change.key] = change
        pending[change.key] = change
    }

    override suspend fun removeFromOutbox(keys: List<ChangeKey>) = mutex.withLock {
        keys.forEach { pending.remove(it) }
    }

    override suspend fun currentHlc(entity: Entity, id: String): String? =
        mutex.withLock { records[ChangeKey(entity, id)]?.hlc }

    override suspend fun applyRemote(change: Change) = mutex.withLock {
        records[change.key] = change
        Unit
    }

    suspend fun record(entity: Entity, id: String): Change? =
        mutex.withLock { records[ChangeKey(entity, id)] }
}
